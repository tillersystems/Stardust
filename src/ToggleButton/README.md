# ToggleButton

### Usage

A Toggle Button represents the switch between two states, on or off.

```jsx
import { ToggleButton } from '@tillersystems/stardust';
```

<!-- STORY -->

<!-- PROPS -->

### Properties

| Name               | Required |   Type   | DefaultValue |                          Description                           |
| ------------------ | :------: | :------: | :----------: | :------------------------------------------------------------: |
| `className`        |    -     | `string` |     `''`     |              className needed by styled component              |
| `isChecked`        |    -     |  `bool`  |     `''`     |  Whether the button is checked or not in controlled component  |
| `isDefaultChecked` |    -     |  `bool`  |   `false`    | Whether the button is checked or not in uncontrolled component |
| `isDisabled`       |    -     |  `bool`  |   `false`    |             Whether the button is disabled or not              |
| `onToggle`         |    -     |  `func`  |  `() => {}`  |               Callback when component is clicked               |

### Example

```jsx
import { ToggleButton } from '@tillersystems/stardust';

render() {
  return <ToggleButton />
}
```
