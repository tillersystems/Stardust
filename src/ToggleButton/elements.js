import styled, { css } from 'styled-components';

const Checkbox = styled.input`
  position: absolute;

  width: 0;

  opacity: 0;
`;

const Toggle = styled.span`
  display: inline-block;

  position: relative;

  width: 3.8rem;
  height: 2rem;

  border-radius: 1rem;

  background: ${({ checked, theme: { palette } }) =>
    checked
      ? `${palette.primary.default} linear-gradient(
          0deg,
          ${palette.whiteOpacity(0)} 0%,
          ${palette.whiteOpacity(0.1)} 100%
        )`
      : palette.lightGrey};

  border: 1px solid
    ${({ checked, theme: { palette } }) => (checked ? palette.primary.default : palette.lightGrey)};

  cursor: pointer;
  pointer-events: none;

  &:before {
    content: '';

    display: block;

    position: absolute;
    top: 0.1rem;
    left: 0.2rem;

    cursor: pointer;

    height: 1.6rem;
    width: 1.6rem;

    border-radius: 1.6rem;

    background: ${({ theme: { palette } }) => palette.white};

    box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.15);

    transition: transform 250ms ease;
    transform: translateX(${({ checked }) => (checked ? '1.7rem' : 0)});
    will-change: transform;
  }

  ${({ readOnly }) =>
    readOnly
      ? css`
          opacity: 0.4;
        `
      : css`
          &:active {
            &:before {
              width: 3rem;
              left: ${p => (p.checked ? '0' : '6px')};
            }
          }
        `};
`;

export { Checkbox, Toggle };
