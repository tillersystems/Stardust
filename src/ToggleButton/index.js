import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

import { Checkbox, Toggle } from './elements';

/**
 * A Toggle Button represents the switch between two states, on or off.
 *
 * @return {jsx}
 */

/* eslint-disable jsx-a11y/click-events-have-key-events */
/* eslint-disable jsx-a11y/interactive-supports-focus */
/* eslint-disable jsx-a11y/no-static-element-interactions */

class ToggleButton extends PureComponent {
  /** Display name. */
  static displayName = 'ToggleButton';

  /** Internal state. */
  state = {
    checked: this.props.checked, // eslint-disable-line react/destructuring-assignment
  };

  componentDidUpdate(prevProps) {
    const { checked } = this.props;

    if (checked !== prevProps.checked) {
      this.setState({ checked });
    }
  }

  /**
   * Handles Toggle of Button
   */
  handleToggle = () => {
    const { disabled, onToggle } = this.props;
    const { checked } = this.state;

    if (!disabled && onToggle) {
      onToggle(!checked);
    } else if (!disabled) {
      this.setState(prevState => ({ checked: !prevState.checked }));
    } else {
      return null;
    }
  };

  render() {
    const { className, disabled } = this.props;
    const { checked } = this.state;

    return (
      <div className={className} onClick={this.handleToggle}>
        <Checkbox type="checkbox" checked={checked} readOnly />
        <Toggle checked={checked} readOnly={disabled} data-testid="toggle-button" />
      </div>
    );
  }
}

/** Prop types. */
ToggleButton.propTypes = {
  /** Whether the button is checked or not */
  checked: PropTypes.bool,

  /** className needed by styled-components */
  className: PropTypes.string,

  /** Whether the button is disabled or not */
  disabled: PropTypes.bool,

  /** Callback when component is clicked */
  onToggle: PropTypes.func,
};

/** Default props. */
ToggleButton.defaultProps = {
  checked: false,
  className: '',
  disabled: false,
  onToggle: null,
};

export default styled(ToggleButton)`
  cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'pointer')};
  display: inline-block;
  height: 2rem;
  position: relative;
  vertical-align: middle;
`;
