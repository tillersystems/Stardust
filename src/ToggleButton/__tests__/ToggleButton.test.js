import React from 'react';
import { fireEvent } from 'react-testing-library';

import ToggleButton from '..';

describe('<ToggleButton />', () => {
  test('should render without a problem', () => {
    const { container } = render(<ToggleButton />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render without a problem when checked', () => {
    const { container } = render(<ToggleButton checked />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render without a problem when disabled', () => {
    const { container, getByTestId } = render(<ToggleButton disabled />);
    const toggleNode = getByTestId('toggle-button');

    expect(container.firstChild).toHaveStyleRule('cursor', 'not-allowed');
    expect(toggleNode).toHaveStyleRule('opacity', '0.4');
  });

  test('should toggle', () => {
    const { container } = render(<ToggleButton />);

    fireEvent.click(container.firstChild);

    expect(render).toMatchSnapshot();

    fireEvent.click(container.firstChild);

    expect(render).toMatchSnapshot();
  });

  test('should call change handler when enabled', () => {
    const handleToggleMock = jest.fn();
    const { container } = render(<ToggleButton checked onToggle={handleToggleMock} />);

    fireEvent.click(container.firstChild);

    expect(handleToggleMock).toHaveBeenCalledTimes(1);
  });

  test('should not call change handler when disabled', () => {
    const handleToggleMock = jest.fn();
    const { container } = render(<ToggleButton checked disabled onToggle={handleToggleMock} />);

    fireEvent.click(container.firstChild);

    expect(handleToggleMock).not.toHaveBeenCalled();
  });
});
