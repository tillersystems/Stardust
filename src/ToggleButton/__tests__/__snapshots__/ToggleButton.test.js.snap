// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ToggleButton /> should render without a problem 1`] = `
.c1 {
  position: absolute;
  width: 0;
  opacity: 0;
}

.c2 {
  display: inline-block;
  position: relative;
  width: 3.8rem;
  height: 2rem;
  border-radius: 1rem;
  background: hsl(216,45%,98%);
  border: 1px solid hsl(216,45%,98%);
  cursor: pointer;
  pointer-events: none;
}

.c2:before {
  content: '';
  display: block;
  position: absolute;
  top: 0.1rem;
  left: 0.2rem;
  cursor: pointer;
  height: 1.6rem;
  width: 1.6rem;
  border-radius: 1.6rem;
  background: hsl(0,100%,100%);
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.15);
  -webkit-transition: -webkit-transform 250ms ease;
  -webkit-transition: transform 250ms ease;
  transition: transform 250ms ease;
  -webkit-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
  will-change: transform;
}

.c2:active:before {
  width: 3rem;
  left: 6px;
}

.c0 {
  display: inline-block;
  position: relative;
  height: 2rem;
  vertical-align: middle;
  cursor: pointer;
}

<Styled(ToggleButton)>
  <ToggleButton
    checked={false}
    className="c0"
    disabled={false}
    onToggle={null}
  >
    <div
      className="c0"
      onClick={[Function]}
    >
      <styled.input
        checked={false}
        readOnly={true}
        type="checkbox"
      >
        <input
          checked={false}
          className="c1"
          readOnly={true}
          type="checkbox"
        />
      </styled.input>
      <styled.span
        checked={false}
        readOnly={false}
      >
        <span
          checked={false}
          className="c2"
          readOnly={false}
        />
      </styled.span>
    </div>
  </ToggleButton>
</Styled(ToggleButton)>
`;

exports[`<ToggleButton /> should render without a problem when checked and disabled 1`] = `
.c1 {
  position: absolute;
  width: 0;
  opacity: 0;
}

.c2 {
  display: inline-block;
  position: relative;
  width: 3.8rem;
  height: 2rem;
  border-radius: 1rem;
  background: hsl(200,74%,46%)linear-gradient(0deg,rgba(255,255,255,0) 0%,rgba(255,255,255,0.1) 100%);
  border: 1px solid hsl(200,74%,46%);
  cursor: pointer;
  pointer-events: none;
  opacity: 0.4;
}

.c2:before {
  content: '';
  display: block;
  position: absolute;
  top: 0.1rem;
  left: 0.2rem;
  cursor: pointer;
  height: 1.6rem;
  width: 1.6rem;
  border-radius: 1.6rem;
  background: hsl(0,100%,100%);
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.15);
  -webkit-transition: -webkit-transform 250ms ease;
  -webkit-transition: transform 250ms ease;
  transition: transform 250ms ease;
  -webkit-transform: translateX(1.7rem);
  -ms-transform: translateX(1.7rem);
  transform: translateX(1.7rem);
  will-change: transform;
}

.c0 {
  display: inline-block;
  position: relative;
  height: 2rem;
  vertical-align: middle;
  cursor: not-allowed;
}

<Styled(ToggleButton)
  checked={true}
  disabled={true}
>
  <ToggleButton
    checked={true}
    className="c0"
    disabled={true}
    onToggle={null}
  >
    <div
      className="c0"
      onClick={[Function]}
    >
      <styled.input
        checked={true}
        readOnly={true}
        type="checkbox"
      >
        <input
          checked={true}
          className="c1"
          readOnly={true}
          type="checkbox"
        />
      </styled.input>
      <styled.span
        checked={true}
        readOnly={true}
      >
        <span
          checked={true}
          className="c2"
          readOnly={true}
        />
      </styled.span>
    </div>
  </ToggleButton>
</Styled(ToggleButton)>
`;

exports[`<ToggleButton /> should render without a problem when checked and not disabled 1`] = `
.c1 {
  position: absolute;
  width: 0;
  opacity: 0;
}

.c2 {
  display: inline-block;
  position: relative;
  width: 3.8rem;
  height: 2rem;
  border-radius: 1rem;
  background: hsl(200,74%,46%)linear-gradient(0deg,rgba(255,255,255,0) 0%,rgba(255,255,255,0.1) 100%);
  border: 1px solid hsl(200,74%,46%);
  cursor: pointer;
  pointer-events: none;
}

.c2:before {
  content: '';
  display: block;
  position: absolute;
  top: 0.1rem;
  left: 0.2rem;
  cursor: pointer;
  height: 1.6rem;
  width: 1.6rem;
  border-radius: 1.6rem;
  background: hsl(0,100%,100%);
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.15);
  -webkit-transition: -webkit-transform 250ms ease;
  -webkit-transition: transform 250ms ease;
  transition: transform 250ms ease;
  -webkit-transform: translateX(1.7rem);
  -ms-transform: translateX(1.7rem);
  transform: translateX(1.7rem);
  will-change: transform;
}

.c2:active:before {
  width: 3rem;
  left: 0;
}

.c0 {
  display: inline-block;
  position: relative;
  height: 2rem;
  vertical-align: middle;
  cursor: pointer;
}

<Styled(ToggleButton)
  checked={true}
  disabled={false}
>
  <ToggleButton
    checked={true}
    className="c0"
    disabled={false}
    onToggle={null}
  >
    <div
      className="c0"
      onClick={[Function]}
    >
      <styled.input
        checked={true}
        readOnly={true}
        type="checkbox"
      >
        <input
          checked={true}
          className="c1"
          readOnly={true}
          type="checkbox"
        />
      </styled.input>
      <styled.span
        checked={true}
        readOnly={false}
      >
        <span
          checked={true}
          className="c2"
          readOnly={false}
        />
      </styled.span>
    </div>
  </ToggleButton>
</Styled(ToggleButton)>
`;

exports[`<ToggleButton /> should render without a problem when unchecked and disabled 1`] = `
.c1 {
  position: absolute;
  width: 0;
  opacity: 0;
}

.c2 {
  display: inline-block;
  position: relative;
  width: 3.8rem;
  height: 2rem;
  border-radius: 1rem;
  background: hsl(216,45%,98%);
  border: 1px solid hsl(216,45%,98%);
  cursor: pointer;
  pointer-events: none;
  opacity: 0.4;
}

.c2:before {
  content: '';
  display: block;
  position: absolute;
  top: 0.1rem;
  left: 0.2rem;
  cursor: pointer;
  height: 1.6rem;
  width: 1.6rem;
  border-radius: 1.6rem;
  background: hsl(0,100%,100%);
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.15);
  -webkit-transition: -webkit-transform 250ms ease;
  -webkit-transition: transform 250ms ease;
  transition: transform 250ms ease;
  -webkit-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
  will-change: transform;
}

.c0 {
  display: inline-block;
  position: relative;
  height: 2rem;
  vertical-align: middle;
  cursor: not-allowed;
}

<Styled(ToggleButton)
  checked={false}
  disabled={true}
>
  <ToggleButton
    checked={false}
    className="c0"
    disabled={true}
    onToggle={null}
  >
    <div
      className="c0"
      onClick={[Function]}
    >
      <styled.input
        checked={false}
        readOnly={true}
        type="checkbox"
      >
        <input
          checked={false}
          className="c1"
          readOnly={true}
          type="checkbox"
        />
      </styled.input>
      <styled.span
        checked={false}
        readOnly={true}
      >
        <span
          checked={false}
          className="c2"
          readOnly={true}
        />
      </styled.span>
    </div>
  </ToggleButton>
</Styled(ToggleButton)>
`;

exports[`<ToggleButton /> should render without a problem when unchecked and not disabled 1`] = `
.c1 {
  position: absolute;
  width: 0;
  opacity: 0;
}

.c2 {
  display: inline-block;
  position: relative;
  width: 3.8rem;
  height: 2rem;
  border-radius: 1rem;
  background: hsl(216,45%,98%);
  border: 1px solid hsl(216,45%,98%);
  cursor: pointer;
  pointer-events: none;
}

.c2:before {
  content: '';
  display: block;
  position: absolute;
  top: 0.1rem;
  left: 0.2rem;
  cursor: pointer;
  height: 1.6rem;
  width: 1.6rem;
  border-radius: 1.6rem;
  background: hsl(0,100%,100%);
  box-shadow: 0 2px 4px 0 rgba(0,0,0,0.15);
  -webkit-transition: -webkit-transform 250ms ease;
  -webkit-transition: transform 250ms ease;
  transition: transform 250ms ease;
  -webkit-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
  will-change: transform;
}

.c2:active:before {
  width: 3rem;
  left: 6px;
}

.c0 {
  display: inline-block;
  position: relative;
  height: 2rem;
  vertical-align: middle;
  cursor: pointer;
}

<Styled(ToggleButton)
  checked={false}
  disabled={false}
>
  <ToggleButton
    checked={false}
    className="c0"
    disabled={false}
    onToggle={null}
  >
    <div
      className="c0"
      onClick={[Function]}
    >
      <styled.input
        checked={false}
        readOnly={true}
        type="checkbox"
      >
        <input
          checked={false}
          className="c1"
          readOnly={true}
          type="checkbox"
        />
      </styled.input>
      <styled.span
        checked={false}
        readOnly={false}
      >
        <span
          checked={false}
          className="c2"
          readOnly={false}
        />
      </styled.span>
    </div>
  </ToggleButton>
</Styled(ToggleButton)>
`;
